Write citations to bibtex

Chris1221 opened the pull request:
I've added an option re: issue #157 (and #144) to convert stored DOIs to `bibtex` on the fly through CrossRef's API and write them to a file within `cli.write_citations`. The way I've done it is pretty straightforward, and hopefully easily understandable. It's a very basic implementation right now, and I'm very open to adding in some more interesting extensions.

If the `bibtex` flag is given with a string, then a connection is opened to that path, the stored DOIs are converted, then written. An internet connection *is required* (and this is checked). The connection overwrites itself each time the program is run, so that we don't have giant files unintentionally.  

If the passed citation is not a valid `stdpopsim.Citation` (or is missing the DOI for some reason) then it fails with a custom `InvalidCitation` exception, and if the DOI is present but the connection to the server fails for whatever reason (either lack of internet access or if the DOI doesn't return a valid page) then it raises a `ConnectionFailed` exception. Hopefully this is sufficiently verbose.

The testing was a little bit more involved but hopefully pretty transparent. I mocked out the `requests.get` function to return a class with the right attributes, so that we can test the cases where

1. Everything works
2. The DOI is missing
3. The connection is bad

I also have a test case for running the whole `cli.write_citations` function, with a mocked out output stream that I test against the mocked out `requests.get`. Lots of mocks. 

I rebased against master, and squashed. Let me know what everyone thinks! 

Some further thoughts:

- I introduced a dependency on `requests` which might not be the greatest idea, given our pretty minimal set of dependencies. If anyone has any thoughts for how we could do this differently and get rid of the dependency, please let me know!
- The patch test has failed, but to my knowledge I tested every line in the diff. A bit confused.

apragsdale commented:
I think that is incredible! I love it

Chris1221 commented:
Thanks @apragsdale! :)

andrewkern commented on tests/test_cli.py:
```diff
@@ -448,12 +448,79 @@ def test_genetic_map(self):
         engine = stdpopsim.get_default_engine()
         stdout, stderr = capture_output(
-                cli.write_citations, engine, model, contig)
+                cli.write_citations, engine, model, contig, None)
```
i'd love it if you used a local variable here to describe the bibtex flag, rather than just pass `None`.  As written it's opaque.

Chris1221 replied:
Perfect, sorry I can see now that it is definitely difficult to parse. I'll fix this while changing it over to `urllib` with @jeromekelleher's suggestions!

Chris1221 commented:
That sounds great @jeromekelleher, I can come in whenever it works for you!

jeromekelleher commented:
> That sounds great @jeromekelleher, I can come in whenever it works for you!

Excellent! Can you drop me an email to arrange please?

Chris1221 commented on stdpopsim/cli.py:
```diff
@@ -217,13 +218,60 @@ def write_output(ts, args):
+def get_bibtex(citation):
+    """Retrieve the bibtex of a citation from Crossref."""
+    try:
+        bib = get(citation.doi, headers={"Accept": "text/bibliography; style=bibtex"})
```
Ah, I do see that now. Sorry about that, as you say this could definitely be easily done in `urllib` instead. I'll see if I can easily convert it over.

jeromekelleher replied:
No need to apologise Chris, this is great work!

Chris1221 commented on stdpopsim/cli.py:
```diff
@@ -217,13 +218,60 @@ def write_output(ts, args):
+        self.status_code = status_code
+
+
+def get_bibtex(citation):
```
Very good point, that would be a much better place for it.

Chris1221 commented:
Brief summary of changes from the above:

- Replaced the dependency on the `requests` library in favour of `urllib.request`
- Moved the `get_bibtex` function to the `stdpopsim.Citation.to_bibtex()` method.
    - Remove the custom exception for malformed citations as a consequence of this
- Add a test for a complete run through from the args down, hitting that last remaining line of code.
- Added a variable to describe the behaviour of `args.bibtex` in the other tests which use `cli.write_citations()`.

Chris1221 commented:
I had to do something funky and mock out `argparse.Filetype.__call__` to correctly test (and basically just test that a connection was being opened to the right file when the function is run). But otherwise everything is the same.

jeromekelleher commented:
Beautiful, thanks @Chris1221! Merging.

agladstein commented:
sorry, but how do you run this?
I tried
```
stdpopsim homsap -c chr22 -o foo.ts 2 --bibtex foo.bib
```

jeromekelleher commented:
That should work --- what are you getting?

agladstein commented:
Ah sorry! I had built the CLI locally - I needed to run it as `python3 -m stdpopsim`

Chris1221 commented:
Also the option should be —bibtex_file. Did you manage to get it working?

agladstein commented:
Oh, huh... `--bibtex foo.bib` worked.

Chris1221 commented:
Oh wait @agladstein, you're right. That's incredibly weird, I'm even able to pass `--bibte` and it will write the bibtex to the file... @jeromekelleher do you know if `argparse` might be doing some sort of partial string matching or something to the argument name? `--bibet` (swapping the t and the e, so that it is not a part of the actual name) correctly fails.

Chris1221 commented:
I can't find any documentation on this in `argparse`, but it seems like it's accepting the minimum unique string to match to the argument?

Examples:

- `--bib` works
- `--b` works (since it is the only argument that starts with b but
    - `-b` fails (because we haven't explicitly defined it)

Also:
- `--help_mod` works for `--help_models`
- So does `--help_m` (as it is unique)
- `--help-` fails with the following message:

```
stdpopsim homsap: error: ambiguous option: --help- could match --help-models, --help-genetic-maps
```

So, I'm inclined to believe this is intended behaviour of the argument parser, but I did have a small amount of panic thinking we had messed it up!

Chris1221 commented:
Sorry, that might have been common knowledge to everyone else, but this is the first time I've seen it!

jeromekelleher commented:
I think this is standard behaviour from argparse @Chris1221, nothing for us to worry about.
